File: solutions/2026-07-20/cpp/2026-07-20-shift-2d-grid.cpp
```cpp
/*
Problem Summary:
Shifts a 2D grid k times according to specific movement rules and returns the modified grid.

Link: https://leetcode.com/problems/shift-2d-grid/

Approach:
The core idea is to simulate the shifting process. Since the grid dimensions and k are relatively small, a direct simulation is feasible.
However, performing k individual shift operations can be inefficient if k is large.
We can observe that the shifting operation is cyclical. The total number of elements in the grid is m * n.
After m * n shifts, the grid will return to its original state. Therefore, we only need to perform k % (m * n) shifts.

To optimize further and avoid repeated shifting, we can calculate the final position of each element directly.
We can flatten the 2D grid into a 1D array conceptually. An element at `grid[i][j]` would be at index `i * n + j` in a 1D representation.
After `k` shifts, an element originally at index `idx` in the 1D array will move to index `(idx + k) % (m * n)`.
We can then convert this new 1D index back to 2D coordinates:
new_row = new_idx / n
new_col = new_idx % n

We create a new grid of the same dimensions and populate it with elements from the original grid placed at their calculated final positions.

Time Complexity:
O(m * n), where m is the number of rows and n is the number of columns. We iterate through the grid once to calculate the new positions and populate the result grid.

Space Complexity:
O(m * n) to store the new grid. If we are allowed to modify the input grid in place, the space complexity could be O(1) if done carefully, but creating a new grid is simpler and safer.
*/

#include <vector>
#include <iostream>

class Solution {
public:
    std::vector<std::vector<int>> shiftGrid(std::vector<std::vector<int>>& grid, int k) {
        // Get dimensions of the grid
        int m = grid.size();
        int n = grid[0].size();

        // Calculate the effective number of shifts
        // The grid returns to its original state after m * n shifts.
        // So, we only need to consider k modulo (m * n).
        int effective_k = k % (m * n);

        // If effective_k is 0, no shifts are needed, return the original grid.
        if (effective_k == 0) {
            return grid;
        }

        // Create a new grid to store the shifted elements.
        // This avoids overwriting elements before they are moved.
        std::vector<std::vector<int>> shifted_grid(m, std::vector<int>(n));

        // Iterate through each element of the original grid
        for (int r = 0; r < m; ++r) {
            for (int c = 0; c < n; ++c) {
                // Calculate the 1D index of the current element
                int current_1d_index = r * n + c;

                // Calculate the new 1D index after k shifts
                int new_1d_index = (current_1d_index + effective_k) % (m * n);

                // Convert the new 1D index back to 2D coordinates
                int new_r = new_1d_index / n;
                int new_c = new_1d_index % n;

                // Place the current element in its new position in the shifted_grid
                shifted_grid[new_r][new_c] = grid[r][c];
            }
        }

        // Return the grid after applying k shifts
        return shifted_grid;
    }
};

/*
// Example Usage (for testing purposes, not part of the LeetCode submission)
int main() {
    Solution sol;

    // Example 1
    std::vector<std::vector<int>> grid1 = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};
    int k1 = 1;
    std::vector<std::vector<int>> result1 = sol.shiftGrid(grid1, k1);
    std::cout << "Example 1 Output:" << std::endl;
    for (const auto& row : result1) {
        for (int val : row) {
            std::cout << val << " ";
        }
        std::cout << std::endl;
    }
    // Expected: [[9,1,2],[3,4,5],[6,7,8]]

    // Example 2
    std::vector<std::vector<int>> grid2 = {{3, 8, 1, 9}, {19, 7, 2, 5}, {4, 6, 11, 10}, {12, 0, 21, 13}};
    int k2 = 4;
    std::vector<std::vector<int>> result2 = sol.shiftGrid(grid2, k2);
    std::cout << "\nExample 2 Output:" << std::endl;
    for (const auto& row : result2) {
        for (int val : row) {
            std::cout << val << " ";
        }
        std::cout << std::endl;
    }
    // Expected: [[12,0,21,13],[3,8,1,9],[19,7,2,5],[4,6,11,10]]

    // Example 3
    std::vector<std::vector<int>> grid3 = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};
    int k3 = 9;
    std::vector<std::vector<int>> result3 = sol.shiftGrid(grid3, k3);
    std::cout << "\nExample 3 Output:" << std::endl;
    for (const auto& row : result3) {
        for (int val : row) {
            std::cout << val << " ";
        }
        std::cout << std::endl;
    }
    // Expected: [[1,2,3],[4,5,6],[7,8,9]]

    return 0;
}
*/
```

File: solutions/2026-07-20/javascript/2026-07-20-shift-2d-grid.js
// Problem: Shift 2D Grid
// Summary: Shifts elements in a 2D grid cyclically to the right and down, k times.
// Link: https://leetcode.com/problems/shift-2d-grid/
//
// Approach:
// The key insight is that the grid shift operation can be thought of as a single
// linear shift if we flatten the 2D grid into a 1D array.
//
// 1. Flatten the 2D grid into a 1D array.
// 2. Calculate the effective number of shifts by taking `k` modulo the total number of elements in the grid. This is because shifting by the total number of elements brings the grid back to its original state.
// 3. Perform the shift on the 1D array. The element at index `i` will move to index `(i + effective_k) % total_elements`.
// 4. Reconstruct the 2D grid from the shifted 1D array.
//
// Time Complexity: O(m * n) where m is the number of rows and n is the number of columns.
// We iterate through the grid once to flatten it, once to perform the shift (conceptually), and once to reconstruct it.
//
// Space Complexity: O(m * n) to store the flattened 1D array.

/**
 * @param {number[][]} grid
 * @param {number} k
 * @return {number[][]}
 */
var shiftGrid = function(grid, k) {
    const m = grid.length;
    const n = grid[0].length;
    const totalElements = m * n;

    // Calculate the effective number of shifts.
    // Shifting by totalElements brings the grid back to its original state.
    const effectiveK = k % totalElements;

    // If effectiveK is 0, no shift is needed, return the original grid.
    if (effectiveK === 0) {
        return grid;
    }

    // Create a 1D array to represent the flattened grid.
    const flatGrid = new Array(totalElements);

    // Flatten the 2D grid into a 1D array.
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            flatGrid[i * n + j] = grid[i][j];
        }
    }

    // Create a new 1D array to store the shifted elements.
    const shiftedFlatGrid = new Array(totalElements);

    // Perform the shift operation on the 1D array.
    // Each element at index `i` moves to `(i + effectiveK) % totalElements`.
    for (let i = 0; i < totalElements; i++) {
        const newIndex = (i + effectiveK) % totalElements;
        shiftedFlatGrid[newIndex] = flatGrid[i];
    }

    // Reconstruct the 2D grid from the shifted 1D array.
    const resultGrid = Array(m).fill(0).map(() => Array(n).fill(0));
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            resultGrid[i][j] = shiftedFlatGrid[i * n + j];
        }
    }

    return resultGrid;
};
